Feature: collect geolocation and other user information

sebsilas opened the pull request:
As discussed in #111 .


Toy app to test:

```r

library(psychTestR)
library(tidyverse)


# Function to flatten a deeply nested list and convert it into a tidy dataframe
unnest_list_to_tidy_df <- function(nested_list) {
  # Helper function to flatten the list recursively
  flatten_list <- function(lst, parent_key = "") {
    out <- list()

    for (name in names(lst)) {
      value <- lst[[name]]
      key <- ifelse(parent_key == "", name, paste0(parent_key, "_", name))

      if (is.list(value) && !is.data.frame(value)) {
        out <- c(out, flatten_list(value, key))
      } else {
        out[[key]] <- as.character(value)
      }
    }

    return(out)
  }

  # Flatten the nested list
  flattened_list <- flatten_list(nested_list)

  # Convert the flattened list to a dataframe
  df <- as_tibble(flattened_list)

  # Convert columns with list elements into separate rows

  tidy_df <- df %>%
    pivot_longer(cols = everything(), names_to = "Variable", values_to = "Value") %>%
    unnest_wider(col = Value, names_sep = "_") %>%
    rename(Value = Value_1) %>%
    filter(!is.na(Value))

  return(tidy_df)
}



make_test(
  elts = join(
    one_button_page("Hi"),

    reactive_page(function(state, ...) {

      session_info <- psychTestR::get_session_info(state, complete = FALSE)

      tb <- shiny::renderTable( {

        user_info <- session_info$user_info

        unnest_list_to_tidy_df(user_info)

      })

      ui <- shiny::tags$div(tb)

      one_button_page(ui)

    }),

    final_page("Bye")
  ),
  opt = test_options(title = "Test get participant info",
                     admin_password = "test",
                     get_user_info = TRUE)
)
```

pmcharrison commented:
Thanks for this, @sebsilas! The diff includes some code you contributed already in your previous MR. It would be good to fix this first. Maybe it will work if you resolve the merge conflicts first?

![image](https://github.com/pmcharrison/psychTestR/assets/22684998/0dbc4709-865e-49b0-af89-89a37d3f2fcd)

sebsilas commented:
Oh, sorry. Try now.

pmcharrison commented on inst/js/get-user-info.js:
```diff
@@ -0,0 +1,36 @@
+    for (var i in navigator) _navigator[i] = navigator[i];
+    delete _navigator.plugins;
+    delete _navigator.mimeTypes;
+    navigatorJSON = JSON.stringify(_navigator);
```
I think I would put the JSONification in the `shiny:sessioninitialized` function instead, I think `getBrowserAndHardwareInfo` should return an ordinary object

pmcharrison commented on inst/js/get-user-info.js:
```diff
@@ -0,0 +1,36 @@
+
+
+// Add an event listener for when the Shiny session is initialized
+$(document).on("shiny:sessioninitialized",function() {
```
missing whitespace

pmcharrison commented on R/server.R:
```diff
@@ -1,5 +1,35 @@
   function(input, output, session) {
+
+    # Get user information
+    user_info <- shiny::reactive({
```
Maybe this reactive should only be executed if `get_user_info` is True?

sebsilas replied:
I did actually have that in and took it out when debugging. Good point!

sebsilas commented:
Hi Peter,

I've implemented your requests in my new commit and also added some new/improved parsing of information. Specifically, now the information is stored as completely unnested. Here's a new test app (you can check with `get_user_info` as `TRUE` and `FALSE`):

```r

library(psychTestR)
library(tidyverse)



make_test(
  elts = join(
    one_button_page("Hi"),

    reactive_page(function(state, ...) {

      session_info <- psychTestR::get_session_info(state, complete = FALSE)

      tb <- shiny::renderTable( {

        user_info <- session_info$user_info

        if(is.null(user_info)) {
          tb <- tibble::tibble(Data = "No user info")
        } else {
          tb <- user_info %>%
            tidyr::pivot_longer(dplyr::everything(), names_to = "Variable", values_to = "Value")
        }

        tb
      })

      ui <- shiny::tags$div(tb)

      one_button_page(ui)

    }),

    final_page("Bye")
  ),
  opt = test_options(title = "Test get participant info",
                     admin_password = "test",
                     get_user_info = TRUE)
)
```

pmcharrison commented on R/ui.R:
```diff
@@ -81,7 +81,9 @@ include_scripts <- function(opt) {
     "js/trigger-button.js",
-    "js/navigate-away.js")
+    "js/navigate-away.js",
+    if(opt$get_user_info) "js/get-user-info.js" else NULL
```
I think this can be shortened to 

```r
if(opt$get_user_info) "js/get-user-info.js"
```

and the NULL will returned automatically if `opt$get_user_info` is `FALSE`

sebsilas replied:
Yup. My style can be a bit overly verbose - I quite like seeing these things explicitly for readability - but I agree this is not necessary and will remove.

pmcharrison commented on R/utils.R:
```diff
@@ -65,3 +65,23 @@ assert_global_is_null <- function(key, state) {
+
+tidy_user_info <- function(user_info) {
+
+  while(purrr::every(user_info, function(x) length(x) == 1)){
```
An alternative to 

```r
  while(purrr::every(user_info, function(x) length(x) == 1)){
    user_info <- purrr::map(user_info, function(x) {
      if(length(x) > 1) return(x) else return(paste0(x, collapse = ", "))
    })
  }
```

would be 

```r
user_info <- map(user_info, jsonlite::toJSON)
```

It's shorter, which is always good, but it also has the advantage that the user can easily reverse the operation to get back vectors by using `jsonlite::fromJSON`. What do you think?

sebsilas replied:
Sure. The only issue with that is that it wraps unnecessary JSON around scalar values and stores the data as Javascript-like, rather than R-like, when we are in R world:

<img width="422" alt="Screenshot 2024-06-25 at 10 06 57" src="https://github.com/pmcharrison/psychTestR/assets/39628712/f69265f5-ecc5-4dc9-a34c-a313ff806a08">


As a hybrid in the middle, how about:

```r
 user_info <- purrr::map(user_info, function(x) {
    if(length(x) > 1) jsonlite::toJSON(x) else x
  })
  ```
  
  This way, we don't store scalar values in the `session` object with unnecessary JSON formatting, but for vectors with length > 1, we can coerce them back into vectors from their JSON string representations. e.g.,:
  
  
<img width="645" alt="Screenshot 2024-06-25 at 10 04 45" src="https://github.com/pmcharrison/psychTestR/assets/39628712/f8ca2914-8d8d-441a-a525-c27b50f73a02">

pmcharrison replied:
Hi @sebsilas, thanks for this! Hmm, good point. You might be able to alternatively address this by passing `auto_unbox = TRUE` to `toJSON`? Documentation below:

> automatically unbox() all atomic vectors of length 1. It is usually safer to avoid this and instead use the unbox() function to unbox individual elements. An exception is that objects of class AsIs (i.e. wrapped in I()) are not automatically unboxed. This is a way to mark single values as length-1 arrays.

Potentially this solves some rough edges in the screenshot you posted (e.g. timeZones is not ideal yet). But your proposal works fine if that doesn't work.

sebsilas replied:
Hm nice idea, although that also doesn't leave some things looking very nice:

<img width="1161" alt="Screenshot 2024-07-01 at 16 45 31" src="https://github.com/pmcharrison/psychTestR/assets/39628712/801872ef-c4fb-40e3-a085-5b811bc38691">



<img width="1071" alt="Screenshot 2024-07-01 at 16 45 16" src="https://github.com/pmcharrison/psychTestR/assets/39628712/6b95940d-345d-4ef1-9b0e-36944d0747bb">


I agree that timeZones isn't ideally structured JSON in my approach, but at least it really does coerce back an R character vector:

```r
test <- rjson::fromJSON('[["Europe/Berlin"],["Europe/Busingen"]]')
```

I'm going to submit a PR with my approach for now, but let me know if you have any other ideas

pmcharrison commented on R/utils.R:
```diff
@@ -65,3 +65,23 @@ assert_global_is_null <- function(key, state) {
+
+  # Tidy names for language specifically
+  idx <- which(grepl("language", names(user_info)))
+  names(user_info)[idx] <- paste0("language", 1:length(idx))
```
Instead of `1:length(idx)` it would be slightly safer to do `seq_along(idx)`, which does not fail if `length(idx) == 0`.

sebsilas replied:
Makes sense!

pmcharrison reviewed (commented):
Looking great, thanks so much! Just added a small number of small remaining comments

pmcharrison reviewed (approved):
Thanks a lot, @sebsilas!
